core/server/proxy/src/test/java/alluxio/StreamCacheTest.java

```diff
@@ -34,11 +36,78 @@ public void operations() throws Exception {
     Assert.assertSame(os, streamCache.invalidate(osId));
     Assert.assertNull(streamCache.invalidate(isId));
     Assert.assertNull(streamCache.invalidate(osId));
+    Assert.assertNull(streamCache.getInStream(isId));
+    Assert.assertNull(streamCache.getOutStream(osId));
     Mockito.verify(is).close();
     Mockito.verify(os).close();
   }
 
   @Test
+  public void concurrentOperations() throws Exception {
+    final StreamCache streamCache = new StreamCache(Constants.HOUR_MS);
+    final FileInStream is = Mockito.mock(FileInStream.class);
+    final FileOutStream os = Mockito.mock(FileOutStream.class);
+
+    // Concurrent put.
+    final AtomicInteger isId = new AtomicInteger(0);
+    Thread putIs = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        isId.set(streamCache.put(is));
+      }
+    });
+    final AtomicInteger osId = new AtomicInteger(0);
+    Thread putOs = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        osId.set(streamCache.put(os));
+      }
+    });
+    putIs.start();
+    putOs.start();
+    putIs.join();
+    putOs.join();
+    Assert.assertSame(is, streamCache.getInStream(isId.get()));
+    Assert.assertSame(os, streamCache.getOutStream(osId.get()));
+
+    // Concurrent get.
+    Thread getIs = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        Assert.assertSame(is, streamCache.getInStream(isId.get()));
+      }
+    });
+    Thread getOs = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        Assert.assertSame(os, streamCache.getOutStream(osId.get()));
+      }
+    });
+    getIs.start();
+    getOs.start();
+    getIs.join();
+    getOs.join();
+
+    // Concurrent get, put, and invalidate.
+    final Integer oldIsId = isId.get();
+    Thread invalidateIs = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        Assert.assertSame(is, streamCache.invalidate(oldIsId));
+      }
+    });
+    putIs.start();
```

Comment:
based on the error you are seeing, it seems like you cannot start the same thread multiple times